File: ios/Vxin/Data/Repositories/GroupRepository.swift
import Foundation

struct RenameGroupBody: Encodable { let name: String }
struct InviteBody: Encodable { let userIds: [String] }
private struct AnnouncementBody: Encodable { let announcement: String }
private struct NicknameBody: Encodable { let nickname: String }
private struct GroupAvatarResponse: Decodable { let avatar: String }
private struct ManageBody: Encodable {
    var mute_all: Bool?
    var no_private_chat: Bool?
    var no_add_friend: Bool?
}
private struct SetRoleBody: Encodable { let role: String }
private struct TransferOwnerBody: Encodable { let userId: String }

/// 群成员管理。与 Android GroupRepository 等价。
final class GroupRepository {
    static let shared = GroupRepository()
    private init() {}

    private let api = APIClient.shared

    func info(_ conversationId: String) async throws -> GroupInfo {
        try await api.send("api/messages/conversation/\(conversationId)/info")
    }

    func rename(_ conversationId: String, name: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)", method: "PUT", body: RenameGroupBody(name: name)
        )
    }

    func setAnnouncement(_ conversationId: String, announcement: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)", method: "PUT", body: AnnouncementBody(announcement: announcement)
        )
    }

    func setNickname(_ conversationId: String, nickname: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/nickname", method: "PUT", body: NicknameBody(nickname: nickname)
        )
    }

    func setAvatar(_ conversationId: String, data: Data, fileName: String) async throws -> String {
        let res: GroupAvatarResponse = try await api.upload(
            "api/messages/conversation/\(conversationId)/avatar", fileData: data, fileName: fileName,
            mimeType: "image/jpeg", fieldName: "avatar", method: "PUT"
        )
        return res.avatar
    }

    func manage(_ conversationId: String, muteAll: Bool? = nil, noPrivateChat: Bool? = nil, noAddFriend: Bool? = nil) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/manage", method: "PUT",
            body: ManageBody(mute_all: muteAll, no_private_chat: noPrivateChat, no_add_friend: noAddFriend)
        )
    }

    func setRole(_ conversationId: String, userId: String, role: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/members/\(userId)/role", method: "PUT", body: SetRoleBody(role: role)
        )
    }

    func transferOwner(_ conversationId: String, userId: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/transfer-owner", method: "POST", body: TransferOwnerBody(userId: userId)
        )
    }

    func invite(_ conversationId: String, userIds: [String]) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/invite", method: "POST", body: InviteBody(userIds: userIds)
        )
    }

    func kick(_ conversationId: String, userId: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/members/\(userId)", method: "DELETE"
        )
    }

    func leave(_ conversationId: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/leave", method: "POST"
        )
    }

    func dissolve(_ conversationId: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/dissolve", method: "POST"
        )
    }

    func qrCode(_ conversationId: String) async throws -> GroupQr {
        try await api.send("api/messages/conversation/\(conversationId)/qr-code")
    }

    func join(token: String) async throws -> JoinGroupResult {
        try await api.send("api/messages/join/\(token)", method: "POST")
    }
}

File: android/app/src/main/java/com/vxin/app/data/repository/GroupRepository.kt
package com.vxin.app.data.repository

import com.vxin.app.data.api.GroupApi
import com.vxin.app.data.model.GroupInfo
import com.vxin.app.data.model.InviteBody
import com.vxin.app.data.model.ManageBody
import com.vxin.app.data.model.NicknameBody
import com.vxin.app.data.model.RenameGroupBody
import com.vxin.app.data.model.SetRoleBody
import com.vxin.app.data.model.UpdateGroupBody
import okhttp3.MultipartBody
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class GroupRepository @Inject constructor(
    private val groupApi: GroupApi,
    socketManager: com.vxin.app.core.realtime.SocketManager,
) {
    /** 群资料/设置/角色/成员变更 → convId */
    val groupChangedEvents = socketManager.groupChangedEvents
    /** 被踢/群解散 → convId */
    val groupGoneEvents = socketManager.groupGoneEvents

    suspend fun info(conversationId: String): GroupInfo = groupApi.info(conversationId)

    suspend fun rename(conversationId: String, name: String) =
        groupApi.rename(conversationId, RenameGroupBody(name))

    suspend fun setAnnouncement(conversationId: String, announcement: String) =
        groupApi.updateInfo(conversationId, UpdateGroupBody(announcement = announcement))

    suspend fun setAvatar(conversationId: String, part: MultipartBody.Part): String =
        groupApi.setAvatar(conversationId, part).avatar

    suspend fun setNickname(conversationId: String, nickname: String) =
        groupApi.setNickname(conversationId, NicknameBody(nickname))

    suspend fun manage(conversationId: String, body: ManageBody) =
        groupApi.manage(conversationId, body)

    suspend fun setRole(conversationId: String, userId: String, role: String) =
        groupApi.setRole(conversationId, userId, SetRoleBody(role))

    suspend fun transferOwner(conversationId: String, userId: String) =
        groupApi.transferOwner(conversationId, com.vxin.app.data.model.TransferOwnerBody(userId))

    suspend fun invite(conversationId: String, userIds: List<String>) =
        groupApi.invite(conversationId, InviteBody(userIds))

    suspend fun kick(conversationId: String, userId: String) =
        groupApi.kick(conversationId, userId)

    suspend fun leave(conversationId: String) = groupApi.leave(conversationId)
    suspend fun dissolve(conversationId: String) = groupApi.dissolve(conversationId)

    suspend fun qrCode(conversationId: String) = groupApi.qrCode(conversationId)

    suspend fun join(token: String) = groupApi.join(token)
}
